File: backend/BookTracker.Core/Entities/Book.cs
namespace BookTracker.Core.Entities;

/// <summary>
/// A book in a user's personal library. Multi-tenancy is enforced by <see cref="UserId"/>;
/// every query in the API layer filters by the authenticated user's id.
/// </summary>
public sealed class Book
{
    public Guid Id { get; set; }

    /// <summary>FK to ASP.NET Identity's AspNetUsers.Id (string).</summary>
    public required string UserId { get; set; }

    public required string Title { get; set; }
    public required string Author { get; set; }

    public string? Isbn { get; set; }
    public string? CoverUrl { get; set; }

    /// <summary>Soft reference to Open Library; null for manual entries. See ADR-0005.</summary>
    public string? OpenLibraryWorkId { get; set; }

    public required BookStatus Status { get; set; }

    /// <summary>Optional 1-5 user rating.</summary>
    public int? Rating { get; set; }

    public string? Notes { get; set; }

    /// <summary>UTC. Set by the server on create.</summary>
    public DateTime DateAdded { get; set; }

    /// <summary>UTC. Required iff <see cref="Status"/> is <see cref="BookStatus.Read"/>; null otherwise.</summary>
    public DateTime? DateFinished { get; set; }
}

File: frontend/src/lib/validation/book.ts
import { z } from "zod";

const bookStatus = z.enum(["wantToRead", "reading", "read"]);

const baseBookFields = {
  title: z.string().min(1).max(300),
  author: z.string().min(1).max(200),
  isbn: z.string().max(20).nullable().optional(),
  coverUrl: z.string().url().max(500).nullable().optional(),
  openLibraryWorkId: z.string().max(50).nullable().optional(),
  status: bookStatus,
  rating: z.number().int().min(1).max(5).nullable().optional(),
  notes: z.string().max(5000).nullable().optional(),
  dateFinished: z.string().datetime().nullable().optional(),
};

export const createBookSchema = z
  .object(baseBookFields)
  .refine(
    (v) =>
      v.status === "read"
        ? v.dateFinished != null
        : v.dateFinished == null,
    {
      message: "dateFinished is required when status is 'Read' and must be null otherwise.",
      path: ["dateFinished"],
    },
  );

export type CreateBookInput = z.infer<typeof createBookSchema>;
